feat(react-router): upgrade to react router 6

ShaneK opened the pull request:
Issue number: resolves #24177

---------

<!-- Please do not submit updates to dependencies unless it fixes an issue. -->

<!-- Please try to limit your pull request to one type (bugfix, feature, etc). Submit multiple pull requests if needed. -->

## What is the current behavior?
Currently, Ionic Framework React Router only supports React Router 5. This has many issues and unsupported/broken features.

## What is the new behavior?
With this change, Ionic Framework will support React Router 6 while still supporting transitions in the same way a native app does.

Most of what caused this change to take a long time is that React Router 5 and React Router 6 have fundamental differences in how they handle components once they're no longer part of the view. In this change, we move away from relying on React Router directly so much and have our own implementation for deciding how views get dealt with during navigation and when they're cleaned up, allowing for us to still transition between them like we need to while still using React Router as much as we possibly can.

This change will also lay the foundation for the migration to React Router 7, which will ideally be easier since most of the hard work has been dealt with here.

## Does this introduce a breaking change?

- [X] Yes
- [ ] No

<!--
  If this introduces a breaking change:
  1. Describe the impact and migration path for existing applications below.
  2. Update the BREAKING.md file with the breaking change.
  3. Add "BREAKING CHANGE: [...]" to the commit description when merging. See https://github.com/ionic-team/ionic-framework/blob/main/docs/CONTRIBUTING.md#footer for more information.
-->


## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

Current dev build (last updated 2026-04-27):

> **⚠️ WARNING:** If you're going to use this dev build on an existing react project, you'll need to migrate to React Router 6. Migrating a large project at this point might be a bad idea since this will not release until v9, which will require further migrations and have other breaking changes! I have a preview of migration documentation for this [here](https://ionic-docs-git-v9-react-router-ionic1.vercel.app/docs/updating/9-0#react-router).

```
8.8.4-dev.11777318673.18d001f6
```

The dev build linked above will be the last one for this branch alone. Everything going forward will be the major-9.0 branch, which may include large breaking changes.

thetaPC committed:
chore(deps): update react router to v6

thetaPC committed:
refactor(utils): update matchPath for rr6

thetaPC committed:
refactor(reactrouterviewstack): update for rr6 and improvements

thetaPC committed:
refactor(test): updated test pages

thetaPC committed:
chore(matchPath): use the correct library

thetaPC committed:
chore(rr6): run lint

thetaPC committed:
chore(StackManager): upgrade to rr6 and add comments

thetaPC committed:
chore(StackManager): more upgrades for rr6

thetaPC committed:
docs(StackManager): update comment

thetaPC committed:
docs(matchPath): remove comment

thetaPC committed:
chore(IonRouteInner): upgrade to rr6

thetaPC committed:
chore(StackManager): run lint

thetaPC committed:
feat(IonRouter): migrate to functional component with react router 6

Co-authored-by: Sean Perkins <13732623+sean-perkins@users.noreply.github.com>

thetaPC committed:
chore(IonRouteInner): update render

thetaPC committed:
docs(IonRouter, ReactRouterViewStack, StackManager): update comments

thetaPC committed:
feat(IonReactRouter): migrate to a functional component and react router 6

Co-authored-by: Sean Perkins <13732623+sean-perkins@users.noreply.github.com>

thetaPC committed:
feat(IonReactMemoryRouter): migrate to a functional component and react router 6

Co-authored-by: Sean Perkins <13732623+sean-perkins@users.noreply.github.com>

thetaPC committed:
chore(IonReactMemoryRouter): remove unused type

thetaPC committed:
feat(IonReactHashRouter): migrate to a functional component and react router 6

Co-authored-by: Sean Perkins <13732623+sean-perkins@users.noreply.github.com>

thetaPC committed:
docs(many): update comments

thetaPC committed:
refactor(many): remove `exact` from test pages

thetaPC committed:
refactor(many): update to use <Navigate>

thetaPC committed:
refactor(many): replace `render` with `element`

thetaPC committed:
refactor(Tabs2): switch to navigate

thetaPC committed:
refactor(many): update to prevent compile errors

thetaPC committed:
refactor(IonReactRouter): split component to use hooks correctly

thetaPC committed:
chore: update package and sync file

thetaPC committed:
docs(IonReactRouter): use updated param names

thetaPC committed:
fix(ReactRouterViewStack, StackManager): use correct value & add value check

thetaPC committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into mh/react-router-6

thetaPC committed:
test(rr5): keep the old pages

thetaPC committed:
test(dynamic-ionpage-classnames): use old version for rr5

thetaPC committed:
chore(rr5, rr6): update package

thetaPC committed:
test(app, routing): use index

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into mh/react-router-6

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into mh/react-router-6

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
WIP

ShaneK committed:
WIP

ShaneK committed:
merge

ShaneK committed:
WIP

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
feat(react-router): have navigation in the routing page generally working really well, lots of work around supporting nested routes and redirection

ShaneK committed:
feat(react-router): adding support for route params

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
fix(react-router): addressing some migration issues with memory and hash router

ShaneK committed:
fix(react-router): improving navigation stack reliability

ShaneK committed:
fix(react-router): improving matching with wildcard and parameter routes

ShaneK committed:
fix(react-router): fixing redirect for test app so it uses replace routing and doesn't trap you, also tests replacement routes

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
chore(test): adding text in refs test for people manually testing

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
fix(react-router): adding support for nested parameter access

ShaneK committed:
fix(react-router): skip IonPage wait for Navigate, adding documentation on progress

ShaneK committed:
trying to fix redirect

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
fix(react-router): fixing dynamic tab and nested outlet tests so they appropriately use wildcard routes

ShaneK committed:
fix(router): correct relative base and param reuse

ShaneK committed:
Fixing test app tabs views

ShaneK committed:
fix(react-router): trying to better handle nested routing

ShaneK committed:
fix(react-router): prevent unmounted Navigate components from triggering redirects

ShaneK committed:
fix(react-router): use state to track IonRouterOutlet ref id in test

ShaneK committed:
fix(react): automatically dismiss inline overlays on navigation

ShaneK committed:
fix(react-router): add wildcard to refs route for nested routing

ShaneK committed:
chore(status): updating status file

ShaneK committed:
chore(react-router): improving continuous testability

ShaneK committed:
fix(react-router): correct tab and nested outlet navigation

ShaneK committed:
fix(react-router): fixing relative paths in tab context tests

ShaneK committed:
fix(react-router): prioritize specific route matches

ShaneK committed:
fix(react-router): prevent incorrect view reuse for parameterized routes

ShaneK committed:
fix(react-router): nested redirect fix

ShaneK committed:
chore(react-router): cleaning up console logs

ShaneK committed:
chore(react-router): removing rr5 test app since it will no longer be able to work, updating workflow to use rr6 tests, updating rr6 status docs

ShaneK committed:
chore(react-router): fixing types, removing usued variable

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
fix(react-router): hide deactivated catch-all routes

ShaneK committed:
fix(react-router): adding legacy peer dep flag during installation to support newer versions of npm until RR6 is released

ShaneK committed:
fix(react-router): fixing views not being cleaned up properly, causing cross navigation issues

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
chore(react-router): code clean up

ShaneK committed:
chore(react-router): refactor

ShaneK committed:
chore(react-router): cleaning up util files

ShaneK committed:
chore(react-router): minor clean up

ShaneK committed:
fix(react): fixing inline overlays being a bit eager to dismiss

ShaneK committed:
fix(overlays): removing old band-aid code

ShaneK committed:
chore(docs): removing status doc link

ShaneK committed:
fix(react-router): preserve nested outlet params when navigating between sibling routes

ShaneK committed:
chore(react-router): adding migration details to BREAKING.md

ShaneK commented on .github/workflows/stencil-nightly.yml:
```diff
@@ -186,7 +186,7 @@ jobs:
       fail-fast: false
       matrix:
-        apps: [reactrouter5]
+        apps: [reactrouter6]
```
I [went ahead and did this in this branch](https://github.com/ionic-team/ionic-framework/pull/30831/commits/fedca460a071e8928deaf6b007e50db41e80966f), figuring it would be an easy enough merge conflict resolution. If you'd like to review in an easier to read way:
https://github.com/ionic-team/ionic-framework/blob/b2a71050344052a4858d0c8b6712f0d5bc9d7219/BREAKING.md

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
fix(react-router): isolate tab history to prevent cross-tab back navigation

thetaPC commented on packages/react-router/test/apps/reactrouter6/package.json:
```diff
@@ -2,21 +2,25 @@
+      "react-router": "$react-router",
+      "react-router-dom": "$react-router-dom"
+    }
+  },
```
Will this be removed once RR6 is released?

thetaPC commented on packages/react-router/test/base/src/pages/routing/OtherPage.tsx:
```diff
@@ -25,7 +23,7 @@ const OtherPage: React.FC<OtherPageProps> = () => {
   return (
     // <IonRouterOutlet id="other" ionPageContainer>
-    //   <Route path="/otherpage" render={() => (
+    //   <Route path="/otherpage" element={
```
After working on this for a bit, do you think it's safe to remove this?

ShaneK replied:
[Done!](https://github.com/ionic-team/ionic-framework/pull/30831/commits/d0451eda090b385ba438f0c61400a16874f1d711#diff-47bed25b7903dd7644b7ca051086937edbaf68237268137ba6922fe7c84ae4adR32)

thetaPC commented on packages/react-router/test/base/src/pages/routing/OtherPage.tsx:
```diff
@@ -39,7 +37,7 @@ const OtherPage: React.FC<OtherPageProps> = () => {
     </IonPage>
-    //   )}></Route>
+    //   }></Route>
     // </IonRouterOutlet>
```
After working on this for a bit, do you think it's safe to remove this?

thetaPC commented on packages/react/src/routing/OutletPageManager.tsx:
```diff
@@ -83,15 +84,16 @@ export class OutletPageManager extends React.Component<OutletPageManagerProps> {
+            <StackManager routeInfo={routeInfo} id={id}>
               <IonRouterOutletInner
                 setRef={(val: HTMLIonRouterOutletElement) => (this.ionRouterOutlet = val)}
+                id={id}
```
Nitpick


```suggestion
                id={id}
                setRef={(val: HTMLIonRouterOutletElement) => (this.ionRouterOutlet = val)}
```

ShaneK commented on packages/react-router/scripts/test_runner.sh:
Yes, I also think it's good for everyone in the same way sync is. It does a lot of the annoying work of building and starting the server and killing the old one for you, which makes iteration much nicer. I 100% intend to commit this to main.

I'll probably create similar scripts for the other packages as we go.

ShaneK committed:
Merge branch 'main' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
Merge branch 'major-9.0' of github.com:ionic-team/ionic-framework into sk/react-router-6

ShaneK committed:
fix(react-router): isolate tab history to prevent cross-tab back navigation (#30854)

Issue number: resolves internal

---------

<!-- Please do not submit updates to dependencies unless it fixes an
issue. -->

<!-- Please try to limit your pull request to one type (bugfix, feature,
etc). Submit multiple pull requests if needed. -->

## What is the current behavior?
When switching between tabs using the tab bar in a React Router 6
application, clicking IonBackButton incorrectly navigates back to the
previous tab. Each tab should have its own isolated navigation history
stack, and the back button should only navigate within the current tab's
history.

## What is the new behavior?

Tab navigation history is now properly isolated. When switching tabs via
the tab bar:
  - The back button only navigates within the current tab's history
- Pressing back on a tab root (with no navigation history) does nothing
instead of navigating to the previous tab or default route
- Within-tab navigation continues to work correctly (e.g., navigating to
a details page and back)
  - Tab history is preserved when switching away and returning to a tab

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

<!--
  If this introduces a breaking change:
1. Describe the impact and migration path for existing applications
below.
  2. Update the BREAKING.md file with the breaking change.
3. Add "BREAKING CHANGE: [...]" to the commit description when merging.
See
https://github.com/ionic-team/ionic-framework/blob/main/docs/CONTRIBUTING.md#footer
for more information.
-->


## Other information

<!-- Any other information that is important to this PR such as
screenshots of how the component looks before and after the change. -->

Current dev build:
```
8.7.12-dev.11765377112.16762e5b
```

ShaneK committed:
chore(git): merge conflict resolution

brandyscarney commented on packages/react/test/base/src/App.tsx:
I am not sure if this intentional behavior, but if you go to http://localhost:3000/dynamic-routes/home locally and then click on `Take me to the newRoute` it navigates to a page that says `New Route Failed`:

<img width="1082" height="642" alt="Image" src="https://github.com/user-attachments/assets/d4648bf8-6748-46e0-85ff-d541f9b0e519" />

A brief look at `packages/react-router/test/base/src/pages/dynamic-routes/DynamicRoutes.tsx` seems to show this as a failed route.

If I click `Add Route` first and then click the link to navigate, it changes the URL but the page doesn't change.

ShaneK replied:
After merging all of my other RR changes into this branch ([07104c2](https://github.com/ionic-team/ionic-framework/pull/30831/commits/07104c2e6b9e6addc833b7313191122c0fafa759)), I seem unable to replicate this. I believe this was fixed indirectly in the change that fixed changes during navigation, because it also fixed an issue with certain navigation directions. If you're able to, please pull and try this one again

brandyscarney commented on BREAKING.md:
```diff
@@ -4,13 +4,76 @@ This is a comprehensive list of the breaking changes introduced in the major ver
+
+For more information on migrating from React Router v5 to v6, refer to the [React Router v6 Upgrade Guide](https://reactrouter.com/en/main/upgrading/v5).
+
 ## Version 8.x
```
Version 8 needs to be moved to its own file, see the `next` branch: https://github.com/ionic-team/ionic-framework/blob/next/BREAKING_ARCHIVE/v8.md

brandyscarney commented on BREAKING.md:
```diff
@@ -4,13 +4,76 @@ This is a comprehensive list of the breaking changes introduced in the major ver
+
+<h2 id="version-9x-framework-specific">Framework Specific</h2>
+
+<h4 id="version-9x-react">React</h4>
```
Should this be React Router?

brandyscarney commented on packages/react/test/base/src/App.tsx:
## Multiple Tabs

Potential bugs:

1. Run the test app locally
2. Navigate to `Multiple Tabs`
3. Open the Menu and go to `Tab 2`
4. Click `Page D`
5. 🐛 See a flicker on the menu toggle button (note: this sometimes happens going from `Page A` to `Page B` also):

https://github.com/user-attachments/assets/e2b117b2-6ab5-4dd5-a661-ccd8633403a7

6. Refresh the page
7. 🐛 See nothing loaded in the view

brandyscarney commented on packages/react/test/base/src/App.tsx:
## Swipe to go back

1. Run the test app locally
2. Navigate to `Swipe to go back`
3. Click on `Details`
4. :bug: See that the test incorrectly uses `ion-buttons` without `slot="start"` resulting in the title wrapping:
   <img width="1082" height="538" alt="Image" src="https://github.com/user-attachments/assets/570e6f9f-7468-4ff2-b685-142c90e3cd42" />

4. Refresh the page.
5. :bug: See that no view is visible.

brandyscarney commented on packages/react-router/test/base/src/pages/swipe-to-go-back/SwipToGoBack.tsx:
Could we rename this file to be spelled correctly? 😆

ShaneK replied:
[jk done lol](https://github.com/ionic-team/ionic-framework/pull/30831/commits/289f6edc676b4bf9e1cdeb4533edac30efa90fbb#diff-6d2032952abdd82e94950731473f25ba69888e5c683b857d31557b6e411222a9)

brandyscarney commented on packages/react-router/scripts/sync.sh:
What was the reason for this update?

brandyscarney commented on packages/react-router/src/ReactRouter/utils/computeParentPath.ts:
```diff
@@ -0,0 +1,259 @@
+      // Find matches at each level, keeping track of the FIRST (shortest) match
+      let firstSpecificMatch: string | undefined = undefined;
+      let firstWildcardMatch: string | undefined = undefined;
+      let indexMatchAtMount: string | undefined = undefined;
```
```suggestion
      let firstSpecificMatch: string | undefined;
      let firstWildcardMatch: string | undefined;
      let indexMatchAtMount: string | undefined;
```

brandyscarney commented on packages/react-router/src/ReactRouter/IonRouter.tsx:
```diff
@@ -1,182 +1,298 @@
-          pathname: location.pathname,
+          ...incomingRouteParams.current,
+          lastPathname: leavingLocationInfo.pathname, // The URL we just came from
+          pathname: location.pathname, // The current (destination) URL
```
Should we change this to something like `prevPathname` and `nextPathname` to make it more obvious?

brandyscarney reviewed (commented):
Looking really good! 🚀 My comments are mostly questions & potential issues with either the test app or the router code.

ShaneK committed:
fix(react-router): support React Router 6 style relative paths in IonRouterOutlet (#30844)

Issue number: resolves [an issue from a
comment](https://github.com/ionic-team/ionic-framework/issues/24177#issuecomment-3624311206)

---------

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying. -->

Currently, Ionic's RR6 doesn't support relative routes in the same way
RR6 does. Routes that do not start with a `/`do not work in the Ionic
RR6 implementation in some cases.

## What is the new behavior?

With this change, we properly support these route styles and more
closely align with normal RR6 route support.

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

<!--
  If this introduces a breaking change:
1. Describe the impact and migration path for existing applications
below.
  2. Update the BREAKING.md file with the breaking change.
3. Add "BREAKING CHANGE: [...]" to the commit description when merging.
See
https://github.com/ionic-team/ionic-framework/blob/main/docs/CONTRIBUTING.md#footer
for more information.
-->


## Other information

Current dev build:
```
 8.7.12-dev.11765307927.1f491e92
```

This PR will be merged into the RR6 branch, which will be squash+merged
into the major 9 branch. This will prevent major 9 from having commits
in the change log on release that reference fixing things that are only
available in major 9.

ShaneK committed:
fix(react-router): moving view stack back to generate ID

ShaneK commented on packages/react-router/src/ReactRouter/ReactRouterViewStack.tsx:
```diff
@@ -1,186 +1,806 @@
+    }
+
+    this.viewItemCounter++;
+    const id = `${outletId}-${this.viewItemCounter}`;
```
At the time, I thought `generateId` was random and I needed to track things for some debugging. I see now that was a mistake, I'll move it back to using `generateId`!

ShaneK replied:
[Done!](https://github.com/ionic-team/ionic-framework/pull/30831/commits/e1e7a2e67a15089b350d9a19fe97d2a4e9da043b)

thetaPC replied:
If using `viewItemCounter` made it easier to debug then we should consider using it with `config.getBoolean('_testing')` like how the animation is [disabled](https://github.com/ionic-team/ionic-framework/blob/098cba4096554dee16632e175ca80b94588117db/core/src/global/ionic-global.ts#L169-L171) during testing.

ShaneK commented on packages/react-router/test/base/src/pages/Main.tsx:
```diff
@@ -9,15 +8,20 @@ import {
       <IonHeader>
         <IonToolbar>
-          <IonTitle>Main</IonTitle>
+          <IonTitle>Test App - React Router v{majorVersion}</IonTitle>
```
I agree with Brandy, but also I don't think this change was necessary. This was very useful when I was side-by-side comparing the RR5 and RR6 test apps to see what functionality was working before vs working now, but we can never actually have multiple RR versions in the same branch, so it doesn't really make sense to keep around now. I'm just going to revert this back to saying "Main"

ShaneK committed:
chore(react-router): lots of comments, code clean up, comment clean up, and a bit of ordering

ShaneK committed:
refactor(react-router): extract helper functions from computeParentPath

ShaneK committed:
refactor(react-router): extract helper functions from render callback

ShaneK commented on packages/react-router/src/ReactRouter/ReactRouterViewStack.tsx:
```diff
@@ -1,186 +1,806 @@
+                outlet: null,
+                matches: contextMatches,
+                isDataRoute: false,
+              };
```
Done! [d44fc11](https://github.com/ionic-team/ionic-framework/pull/30831/commits/d44fc1168b2591e49ca193607ae1e1ef28ad8c70)

ShaneK commented on packages/react-router/test/base/scripts/sync.sh:
```diff
@@ -15,4 +15,4 @@ npm pack ../../../../react
 
 # Install Dependencies
-npm install *.tgz --no-save
+npm install *.tgz --no-save --legacy-peer-deps
```
[Done!](https://github.com/ionic-team/ionic-framework/pull/30831/commits/d0451eda090b385ba438f0c61400a16874f1d711#diff-61e5045cabdf388ab928c32df96a1b37115c90e96f9c199fc58a5b7272952232R18)

ShaneK commented on packages/react-router/test/base/tests/e2e/support/commands.js:
```diff
@@ -82,6 +82,7 @@ Cypress.Commands.add('ionMenuNav', (contains) => {
   // cy.get('ion-menu.show-menu').should('exist');
   // cy.wait(1000)
```
[Done](https://github.com/ionic-team/ionic-framework/pull/30831/commits/d0451eda090b385ba438f0c61400a16874f1d711#diff-c3c3a438c2852ccc28850f84fdf35f46d997e143b474ac84aaab9010cfc95808L82)

ShaneK commented on packages/react/src/routing/OutletPageManager.tsx:
```diff
@@ -83,15 +84,16 @@ export class OutletPageManager extends React.Component<OutletPageManagerProps> {
           this.ionLifeCycleContext = context;
           return (
-            <StackManager routeInfo={routeInfo}>
+            <StackManager routeInfo={routeInfo} id={id}>
```
[Done](https://github.com/ionic-team/ionic-framework/pull/30831/commits/d0451eda090b385ba438f0c61400a16874f1d711#diff-651fdf6e8f3d14b9ee17c1a6a1930696c8c03b93525b6e26bcd2d386dbe531dfL93)

ShaneK committed:
chore(react-router): documentation, comment clean up, and more

ShaneK commented on packages/react-router/test/base/src/pages/routing/Tab2.tsx:
```diff
@@ -51,10 +51,10 @@ const Tab2: React.FC = () => {
           }}
         >
-          Details with Unmount via history.push
+          Details with Unmount via navigate
```
[Cleaned up, good catch](https://github.com/ionic-team/ionic-framework/pull/30831/commits/289f6edc676b4bf9e1cdeb4533edac30efa90fbb#diff-80ce5afff0ae7cabc6e4a67833953a370fdf5bad5e63f472ff64f9bbea1c0918L57)

ShaneK committed:
fix(react-router): cleanup orphaned sibling views after replace navigation

ShaneK committed:
fix(react-router): prevent menu button flicker during non-animated page transitions

ShaneK commented on packages/react/test/base/src/App.tsx:
On my latest commit ([ffa2e06](https://github.com/ionic-team/ionic-framework/pull/30831/commits/ffa2e06b62bcf882c189223a42600faaa5ca94e9)), I was unable to reproduce this. This may have been related to some animation issues that resolved.

brandyscarney replied:
After discussing with Shane the console logs were happening before I clicked the button and the reason the button was not doing anything is because the transition utility sets `pointer-events: none` on both views until the animation is complete. This is most likely just the result of the animation taking a bit longer sometimes. Marking as resolved!

ShaneK committed:
fix(react): preserve framework classes when IonPage className prop changes

ShaneK committed:
test(react-router): fix test for nested outlet rendering on direct navigation

ShaneK committed:
fix(react-router): cleanup sibling views on push navigation to container routes

brandyscarney reviewed (approved):
Woooooooo! 🚀

thetaPC commented on packages/react-router/test/base/src/pages/tab-context/TabContext.tsx:
Found an issue where the active styles are not matching when using the buttons. Notice in the video that both tab buttons have active styles.

https://github.com/user-attachments/assets/cbe5937a-e4ce-4aef-b9ca-55a37ec403a3

thetaPC commented on packages/react-router/test/base/src/pages/nested-outlet/NestedOutlet2.tsx:
Found an issue that the URL and page doesn't match when also using the browser back buttons. My example uses list #1 but it also happens for any list item.

https://github.com/user-attachments/assets/c82452fd-22c2-4192-9e86-a20d50510b40

thetaPC commented on packages/react-router/test/base/src/pages/params/Params.tsx:
Pages and URLs do not match when using the browser's buttons.

https://github.com/user-attachments/assets/077d7f38-f190-460d-8bd5-67a2d385cc30

thetaPC commented on packages/react-router/test/base/src/pages/nested-params/NestedParams.tsx:
Found an issue with the toolbar overlapping when using the browser's buttons.

https://github.com/user-attachments/assets/1d59dc1f-afdc-4132-a865-119800e77eb5

thetaPC commented on packages/react-router/test/base/src/pages/root-splat-tabs/RootSplatTabs.tsx:
Found an issue with active styles on the tab buttons when navigating from root.

https://github.com/user-attachments/assets/20218b6f-dbf2-4ed6-979f-3718a3d5c28c

ShaneK committed:
fix(react-router): prevent white flash during tab switches with selectTab

ShaneK committed:
chore(replace test): adding note to tell manual testers what's expected

ShaneK commented on packages/react-router/test/base/src/pages/replace-action/Replace.tsx:
Good idea, done! [54979f9](https://github.com/ionic-team/ionic-framework/pull/30831/commits/54979f9de8db64ed6757c9e8d9c30eb9dbd05875)

ShaneK committed:
chore(classnames test): adding note explaining the purpose of this test

ShaneK commented on packages/react-router/test/base/src/pages/dynamic-ionpage-classnames/DynamicIonpageClassnames.tsx:
[210331f](https://github.com/ionic-team/ionic-framework/pull/30831/commits/210331f8086eb09ee48c4fc5a92ac005faa0040b)
The point of this page is to make sure that classes added by the framework stick around when they're changed dynamically (e.g. ion-page and can-go-back if you didn't load the page directly but navigated to it from `/`). Brandy actually found a bug thanks to this page in an earlier review.

ShaneK committed:
chore(tests): fixing issue with dynamic classname test

ShaneK committed:
refactor(react): remove IonRedirect and update BREAKING.md for React Router v6

ShaneK committed:
chore(git): merging from main, v8.8.0

ShaneK committed:
fix(react-router): hide all views in outlet for container route transitions, allowing navigation to pages not properly wrapped in ion-pages

ShaneK committed:
fix(react-router): distinguish browser forward from back in POP events

ShaneK committed:
fix(react-router): use location keys for forward detection and prevent overlap in nested outlets

ShaneK committed:
fix(tabs-similar-prefixes): migrating to rr6

ShaneK committed:
fix(react-router): include search params in navigation change detection

ShaneK committed:
fix(react-router): fix dead code in handleNavigateBack breaking browser forward button

ShaneK committed:
fix(react-router): forward index and caseSensitive props through IonRoute

ShaneK committed:
test(nested-params): adding extra coverage for deeply nested routing

ShaneK committed:
fix(react-router): use stored parent paths for index route matching in findViewItemByPath

ShaneK committed:
fix(react-router): fix wildcard route reuse preventing navigation to specific routes after back

ShaneK committed:
fix(react-router): push to forwardStack on non-linear POP so browser forward works

ShaneK committed:
fix(react-router): use reRender callback in getChildrenToRender to clean up stale views

ShaneK committed:
fix(react-router): prevent parameterized routes from catching multi-segment paths meant for wildcards

ShaneK committed:
fix(react-router): add useCallback to HashRouter and MemoryRouter for consistency with BrowserRouter

ShaneK committed:
fix(react-router): cancel in-flight rAFs and MutationObserver on StackManager unmount

ShaneK committed:
fix(react-router): sort fallback params by specificity to prevent less-specific overwrites

ShaneK committed:
fix(react-router): return '/' instead of '' for pathnameBase at root in relative path matching

ShaneK committed:
fix(react-router): render relative catch-all * routes in nested IonRouterOutlet

ShaneK committed:
fix(react-router): cleaning up tab navigation to use proper navigation step post debugging

ShaneK committed:
test(react-router): add coverage for index route reuse across nested outlet tabs

ShaneK committed:
fix(react-router): prevent tail-slice false positives for parameterized routes

ShaneK committed:
test(react-router): add swipe-to-go-back abort coverage for multi-level navigation

ShaneK committed:
fix(react-router): replacing a lot of our custom logic with internal RR6 APIs to do the heavy lifting and fix routing bugs

ShaneK committed:
fix(react-router): using rr6's match routes to prevent having our own custom route matching logic

ShaneK committed:
fix(react-router): guarding against potential change of UNSAFE APIs)

ShaneK committed:
test(react-router): trying to make tests more reliable

ShaneK committed:
fix(react-router): code cleanup and consistency

ShaneK committed:
fix(core): clean up aria-hidden when overlays are removed without dismiss

ShaneK committed:
test(react-router): validating params get passed through redirects

ShaneK committed:
fix(react-router): show correct view for navigate(-n) where n > 1

ShaneK committed:
fix(react-router): set pushedByRoute for routerDirection=none so back button uses history instead of defaultHref

ShaneK committed:
fix(react): preserve query params on IonTabButton href during tab navigation

ShaneK committed:
fix(react-router): preventing content that's transitioning out from being removed before it's supposed to be, fixing animations

ShaneK committed:
test(react-router): adding playwright test structure, creating animation tests in playwright

ShaneK committed:
fix(react-router): sync URL with display on browser back after tab switch

ShaneK committed:
fix(react): synchronous outlet cleanup on unmount

ShaneK committed:
fix(react-router): clean up replaced views with parameterized routes

ShaneK committed:
fix(react-router): prevent swipe-back gesture from starting when entering and leaving views resolve to the same parameterized route

ShaneK committed:
fix(react-router): fire ionViewDidLeave on tab child pages when navigating away from tabs

ShaneK committed:
Fixing test description

ShaneK committed:
fix(react-router): allow defaultHref navigation on direct deep-link to tab child pages

ShaneK committed:
fix(react): fixing issue where modifier keys were blocked while clicking links

ShaneK committed:
fix(react-router): use current route as leaving location for all navigation actions

ShaneK committed:
feat(react-router): add navigateRoot to useIonRouter

ShaneK committed:
fix(react-router): removing previously required wait on tab click tests

ShaneK committed:
fix(react-router): prevent stale animated commit from hiding page after concurrent tab switch

ShaneK committed:
test(react-router): add regression test for catch-all redirect with no ionPageElement

ShaneK committed:
fix(react-router): defer componentDidMount page transition to microtask for React 19 Suspense compatibility

ShaneK committed:
test(react-router): add React 19 test variant for multi-version CI coverage, verified known react 19 issue is caught by it

ShaneK committed:
chore(git): merge

ShaneK committed:
chore(test): fixing test runner core setup

ShaneK committed:
fix(react-router): prevent stale StackManager from destroying views on repeated out-of-scope transitions

ShaneK committed:
fix(react-router): prevent duplicate history entries when replacing to previous URL

ShaneK committed:
chore(test): stylizing manual testing, adding descriptions that try to tell you what each test route is for

ShaneK committed:
fix(react-router): defer nested outlet unmount and prevent forward flash

ShaneK committed:
chore(git): merge

ShaneK committed:
fix(test): add missing slot=start on IonButtons in details test pages

ShaneK committed:
fix(react-router): maintain can-go-back class on nested outlet re-navigation

ShaneK committed:
fix(react): guard against strict mode double-mount in OutletPageManager

ShaneK committed:
fix(react): prevent strict mode double-mount from re-adding ion-page-invisible

ShaneK committed:
fix(react-router): preserve nested ionPage views during parent swipe-back

ShaneK committed:
fix(react-router): preserve non-parameterized views on browser back to retain React state

ShaneK committed:
fix(react-router): scope push cleanup to explicitly preserved views

ShaneK committed:
test(react-router): wait for forward transition to settle before swipe-back assertion

thetaPC reviewed (approved):
LGTM!!!! AMAZING JOB!

ShaneK committed:
fix(react-router): guard delayed unmount against re-entered views

ShaneK committed:
chore(git): merge

ShaneK committed:
fix(stencil): pinning react output target to 1.5.0 until regression is fixed

brandyscarney reviewed (approved):
🎉 🎉 🎉
